src/aspire/operators/diag_matrix.py

```diff
@@ -0,0 +1,622 @@
+"""
+Implements operations for diagonal matrices as used by ASPIRE.
+"""
+
+import numpy as np
+
+from .blk_diag_matrix import BlkDiagMatrix, is_scalar_type
+
+
+class DiagMatrix:
+    """
+    Implements operations for diagonal matrices as used by ASPIRE.
+
+    Currently `DiagMatrix` is implemented as the equivalent of square
+    matrices.  In the future it can be extended to support
+    applications with rectangular shapes.
+    """
+
+    # Developers' Note:
+    # All instances of this class should have priority over ndarray ops
+    #   because we implement them here ourselves.
+    # This is a more np current implementation of __array_priority__
+    #   operator precedence schedule.
+    # Mainly this effects rmul, eg:
+    # np_ary @ diag_mat
+    __array_ufunc__ = None
+
+    def __init__(self, data, dtype=None):
+        """
+        Instantiate a `DiagMatrix` with Numpy `data` shaped (...., self.count),
+        where `self.count` is the length of one diagonal vector.
+
+        Slower axes (if present) are taken to be stack axes.
+
+        :param data: Diagonal matrix entries.
+        :param dtype: Datatype. Default of `None` will attempt
+            passthrough of `data.dtype`.  When explicitly provided, will
+            attempt casting `data` as needed.
+        :return: `DiagMatrix` instance.
+        """
+
+        # Assign the datatype.
+        if dtype is None:
+            dtype = data.dtype
+        self.dtype = np.dtype(dtype)
+
+        # Assign the `data`
+        self._data = data.astype(self.dtype, copy=False)
+
+        # Assign shapes from `data`
+        self.count = self._data.shape[-1]
+        self.stack_shape = self._data.shape[:-1]
+        self.shape = self._data.shape
```

Comment:
Don't we want this to give the shape of the simulated array, i.e. `self._data.shape[:-1] + self._data.shape[-1:] * 2`?

Reply:
I don't think so. It's not used at all so if it is going to be confusing or a point of contention I would prefer just to remove it.

We don't have such a thing for block diagonal.

Reply:
Removed, replaced the only occurances (in unit test) with ._data.shape.

Reply:
That works.